Cython/Compiler/ExprNodes.py

```diff
@@ -11364,13 +11364,8 @@ def compute_c_result_type(self, type1, type2):
                 self, type1, type2)
 
     def py_operation_function(self, code):
-        is_unicode_concat = False
-        if isinstance(self.operand1, FormattedValueNode) or isinstance(self.operand2, FormattedValueNode):
-            is_unicode_concat = True
```

Comment:
I wonder if removing this special case can lead to regressions. Doesn't seem very likely, given that f-strings are always built from unicode string parts. But I don't remember all the optimisations that apply to these parts along the way. Not all of them may be aware of the concrete result type in order to preserve it along the way.

Reply:
As far as I could tell the result type of `FormattedValueNode` was permanently set to `unicode_type` anyway, so it seemed fairly safe. Also, it'd be (probably fairly small?) performance regression rather than than anything breaking (`PyNumber_Add` does work). I can definitely put it back in if wanted, but it seems more likely to cause bugs then fix them.

Reply:
> seems more likely to cause bugs then fix them.

`FormattedValueNode` should only originate from f-strings, and they always produce string parts that are then joined in one way or another. That seems safe. The question is just if the new "unicode or str" decision below succeeds for all of those parts, including operands that are no longer `FormattedValueNode`s themselves (but return a unicode string at runtime).

This PR seems like a fix that should go into 0.29.x, really, in which case I'd like to reduce the risk of regressions a bit more than strictly necessary. I don't think it hurts to keep the `isinstance()` tests, at least in that branch.